### PR TITLE
fix(netcheck): Integrate https and icmp probes in probeplan

### LINKED ISSUE #1138: netcheck: ping and https probes only needed when udp fails
The http and ping probes are only used to try and get a derp latency when udp is fully blocked, i.e. STUN failed entirely.  They should only fire off once we know the udp connectivity fails.

It looks like this was changed a little during the implementation, so check through this logic and make sure the probe plans result in the right probes being executed.

### PR DESCRIPTION
**Pull request by flub**
## Description

The HTTPS and ICMP probes only need to run after the STUN probes have
failed.  This is done in the probe plan by sequencing them after the
stun probes.  If the STUN probes succeed the probe_would_help logic
will shortcut execution.

Because this was all constantly looking up DerpNodes in the DerpMap
they are now created once and put on the heap when the plan is
created.  This means the ProbeReports now carry enough data that they
don't need looking up yet again.

Closes #1138 

## Notes & open questions

~~Because we only have one derp server currently, we never reach enough regions to abort all the probes.  This means we end up waiting for the start of the icmp & http probes.  Once they start the probe-would-help check however aborts the probe sets.~~

~~I'm not really happy with that, it would be better if the probesets for a region would be aborted right away once the region is done.  This ties in with the other refactor of moving the probe-would-help logic upside down that I've already mentioned (see https://github.com/n0-computer/iroh/pull/1180#pullrequestreview-1518251304).~~

~~It's probably fine for now if we create an issue for it?~~

The solution is enough enough: https://github.com/n0-computer/iroh/pull/1220/commits/3bd01082eba46dc62c825ab40a418a9d47db9895

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.

### CODE REVIEW

File: iroh-net/src/hp/netcheck/reportgen.rs [hunk 1]
```diff
@@ -788,35 +788,29 @@ async fn run_probe(
...
+                    Ok(Ok(latency)) => {
+                        result.delay = Some(latency);
                         result.ipv4_can_send = true;
                         result.icmpv4 = true;
                     }
-                    Err(err) => {
-                        warn!("icmp latency measurement failed: {:?}", err);
-                    }
+                    Ok(Err(err)) => warn!("icmp latency measurement failed: {:#}", err),
+                    Err(elapsed) => warn!("icmp latency measurement failed: {:#}", elapsed),
```
Comment (dignifiedquire) [inline iroh-net/src/hp/netcheck/reportgen.rs:805]: maybe call this `measurement timed out`?

File: iroh-net/src/hp/netcheck/reportgen/probes.rs [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+/// compile-time checked but that introduces a giant mess of generics and traits and
+/// associated exploding types.
+///
+/// A [`ProbeSet`] implements [`IntoIterator`] similar to how [`Vec`] does.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub(super) struct ProbeSet {
+    /// Name of this probe set, informational without uniqueness guarantee.
+    name: String,
+    /// The [`ProbeProto`] all the probes in this set have.
+    proto: ProbeProto,
```
Comment (dignifiedquire) [inline iroh-net/src/hp/netcheck/reportgen/probes.rs:140]: I still hate this name..(not your fault)

File: iroh-net/src/hp/netcheck/reportgen/probes.rs [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+            }
+            let retransmit_delay = last_report
+                .region_latency
+                .get(reg.region_id)
+                .map(|l| l * 120 / 100) // increases latency by 20%, why?
+                .unwrap_or(DEFAULT_ACTIVE_RETRANSMIT_DELAY);
+
+            let mut stun_ipv4_probes = ProbeSet::new(reg.region_id, ProbeProto::StunIpv4);
+            let mut stun_ipv6_probes = ProbeSet::new(reg.region_id, ProbeProto::StunIpv6);
+            // let mut https = Vec::new();
```
Comment (dignifiedquire) [inline iroh-net/src/hp/netcheck/reportgen/probes.rs:324]: what about this?

File: iroh-net/src/hp/netcheck/reportgen.rs (hunk 1 above)
Comment (flub) [reply iroh-net/src/hp/netcheck/reportgen.rs:805]: 47e8ba819e2b9f29a5601c4b913a94e4648e2331

File: iroh-net/src/hp/netcheck/reportgen.rs [hunk 4]
```diff
@@ -926,42 +920,99 @@ async fn check_captive_portal(dm: &DerpMap, preferred_derp: Option<u16>) -> Resu
...
-        .with_context(|| format!("no address for node {}", node.name))?;
+    match n.url.host() {
+        Some(url::Host::Domain(hostname)) => {
+            async move {
+                debug!(?proto, %hostname, "Performing DNS lookup for derp addr");
+
+                if let Ok(addrs) = DNS_RESOLVER.lookup_ip(hostname).await {
+                    for addr in addrs {
+                        if addr.is_ipv4() && proto == ProbeProto::StunIpv4 {
+                            let addr = ip::to_canonical(addr);
```
[the remark this answers is not in the document]
Comment (flub) [reply iroh-net/src/hp/netcheck/reportgen.rs:970]: that's copy-pasted code... but yes, totally

79cb1b2a

File: iroh-net/src/hp/netcheck/reportgen/probes.rs (hunk 2 above)
Comment (flub) [reply iroh-net/src/hp/netcheck/reportgen/probes.rs:140]: heh, I did get used to it a little.

File: iroh-net/src/hp/netcheck/reportgen/probes.rs (hunk 3 above)
Comment (flub) [reply iroh-net/src/hp/netcheck/reportgen/probes.rs:324]: just forgotten

d8a04128

File: iroh-net/src/hp/netcheck/reportgen/probes.rs [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+        Self {
+            probes: Vec::new(),
+            name,
+            proto,
+        }
+    }
+
+    fn push(&mut self, probe: Probe) -> Result<()> {
+        if probe.proto() != self.proto {
+            bail!("mismatching probe proto");
```
[the remark this answers is not in the document]
Comment (flub) [reply iroh-net/src/hp/netcheck/reportgen/probes.rs:157]: 3a631cd2

File: iroh-net/src/hp/netcheck/reportgen/probes.rs [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+                    },
+                ],
+            })
+        }
+
+        println!("expected:");
+        println!("{expected_plan}");
+        println!("actual:");
+        println!("{plan}");
+        // TODO: use a fancy diff library, for now just eyeball the output from above...
```
[the remark this answers is not in the document]
Comment (flub) [reply iroh-net/src/hp/netcheck/reportgen/probes.rs:720]: nice

89ab0746

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**